Add never_split signature

tingkai-zhang opened the pull request:
Add ``never_split`` signature to the initialization of ``BertBasicWordSplitter``

tingkai-zhang commented:
Hi, I found the ``never_split`` signature is missing in ``BertBasicWordSplitter``.
It is useful when we have some special cases that we don't want to split.
Or I miss that there is some particular reason of not adding this signature?

tingkai-zhang commented:
I notice that the ``BasicTokenizer`` imported does not has the same signature of what pytorch-bert [master has](https://github.com/huggingface/pytorch-pretrained-BERT/blob/0a9d7c7edb20a3e82cfbb4b72515575543784823/pytorch_pretrained_bert/tokenization.py#L163).

joelgrus commented:
it is possible, but it's not ideal, and I'm pretty sure they'd be happy to do a new release if we asked them, so I'd prefer that solution

joelgrus commented:
https://github.com/huggingface/pytorch-pretrained-BERT/issues/245

WrRan commented on allennlp/data/tokenizers/word_splitter.py:
```diff
@@ -189,8 +189,10 @@ class BertBasicWordSplitter(WordSplitter):
-        self.basic_tokenizer = BertTokenizer(do_lower_case)
+    def __init__(self,
+                 do_lower_case: bool = True,
+                 never_split: Set[str] = ("[UNK]", "[SEP]", "[PAD]", "[CLS]", "[MASK]")) -> None:
```
Should it be `never_split: Tuple[str]` instead of `never_split: Set[str]`?

tingkai-zhang replied:
Yeah, you are right. My mistake.
I was thinking to make it `Set` because it should not be duplicate elements in it.
But I made to be Tuple here just to make it consistent with the original.
Also, since huggingface released their v0.5 pre-trained-bert, do you find there is any change we should make in Allennlp?

WrRan commented:
Hey, @tingkai-zhang .

I think those following resources may be helpful.
1. Joel Grus has upgraded `pytorch-pretrained-bert` to v0.5.0 in #2495. You can merge it to your branch.
2. Add some test cases for the parameter `never_split`. My codebase may be helpful, see https://github.com/WrRan/allennlp/commit/e49ca6c8cc3f761349f150a8022b521c89e00780.

tingkai-zhang commented:
> Hey, @tingkai-zhang .
> 
> I think those following resources may be helpful.
> 
> 1. Joel Grus has upgraded `pytorch-pretrained-bert` to v0.5.0 in #2495. You can merge it to your branch.
> 2. Add some test cases for the parameter `never_split`. My codebase may be helpful, see [WrRan@e49ca6c](https://github.com/WrRan/allennlp/commit/e49ca6c8cc3f761349f150a8022b521c89e00780).

Thank you! Will work on that this weekend.

In your code, do you think `never_split_file_path` is necessary?

Joelgrus said before that it is better to use jsonnet import the external variable. 
#2335

WrRan commented:
@tingkai-zhang Yes. It is a better idea to use `jsonnet` instead of `never_split_file_path`. :+1: 

Another thing. I think it is necessary to add parameter `never_lowercase` for the class `WordpieceIndexer` if we want to correctly `never_split` some text in AllenNlp.

More info please see https://github.com/WrRan/allennlp/blob/e49ca6c8cc3f761349f150a8022b521c89e00780/allennlp/data/token_indexers/wordpiece_indexer.py#L51-L53

tingkai-zhang commented:
Hi, @WrRan ! Sorry for the late!

I have updated the pull request and would you like to take a look at it?
Appreciate your help!

Also, I have some newbie questions.
I find when I clicked green update branch button(which pull my master branch to this), it will print a log in this pull request thread, saying `Merge branch 'master' into patch-1`

However, since I modified two files and committed twice with summaries in my own master branch, I wanna this thread also have two commits, each indicating what file I modified and additional information, just like what my own branch have.

How can I achieve this properly? 

I know these are minor changes but I do want to make my contribution better and follow good conventions! 

Thank you!

tingkai-zhang commented:
Hi, @nelson-liu  !
Sorry to bother you but could you help me on the following issue?
I know It's better to run tests locally as suggested by contribution page but when I ran tests on my branch it gave me 248 errors.

One part of errors likes like this 
`
_________________________________________________ ERROR collecting allennlp/tests/training/momentum_schedulers/inverted_triangular_test.py __________________________________________________
../anaconda3/envs/torch/lib/python3.7/site-packages/py/_path/local.py:668: in pyimport
    __import__(modname)
allennlp/__init__.py:18: in <module>
    import spacy, torch, numpy  # pylint: disable=multiple-imports
../anaconda3/envs/torch/lib/python3.7/site-packages/spacy/__init__.py:4: in <module>
    from .cli.info import info as cli_info
../anaconda3/envs/torch/lib/python3.7/site-packages/spacy/cli/__init__.py:1: in <module>
    from .download import download
../anaconda3/envs/torch/lib/python3.7/site-packages/spacy/cli/download.py:5: in <module>
    import requests
../anaconda3/envs/torch/lib/python3.7/site-packages/requests/__init__.py:115: in <module>
    from .api import request, get, head, post, patch, put, delete, options
../anaconda3/envs/torch/lib/python3.7/site-packages/requests/api.py:13: in <module>
    from . import sessions
../anaconda3/envs/torch/lib/python3.7/site-packages/requests/sessions.py:27: in <module>
    from .adapters import HTTPAdapter
../anaconda3/envs/torch/lib/python3.7/site-packages/requests/adapters.py:44: in <module>
    from urllib3.contrib.socks import SOCKSProxyManager
../anaconda3/envs/torch/lib/python3.7/site-packages/urllib3/contrib/socks.py:27: in <module>
    import socks
../anaconda3/envs/torch/lib/python3.7/site-packages/socks.py:58: in <module>
    from collections import Callable
<frozen importlib._bootstrap>:1032: in _handle_fromlist
    ???
../anaconda3/envs/torch/lib/python3.7/collections/__init__.py:52: in __getattr__
    DeprecationWarning, stacklevel=2)
E   DeprecationWarning: Using or importing the ABCs from 'collections' instead of from 'collections.abc' is deprecated, and in 3.8 it will stop working
===Flaky Test Report===
`

I was wondering is there anything I should check before I run the tests? 
I appreciate your time and help!

Thank you!

nelson-liu commented:
hi @tingkai-zhang , is that the whole traceback? if you could copy and paste the entire terminal output and put it on a gist (gist.github.com) and post it here, i would be able to better assist.

also fyi, most folks are working toward the ACL deadline on Monday, so we might be slow to respond.

tingkai-zhang commented:
> hi @tingkai-zhang , is that the whole traceback? if you could copy and paste the entire terminal output and put it on a gist (gist.github.com) and post it here, i would be able to better assist.
> 
> also fyi, most folks are working toward the ACL deadline on Monday, so we might be slow to respond.

Thanks for replying!
I have created the [gist ](https://gist.github.com/tingkai-zhang/5ea2fef9973af97e0106d28086450e64)here.

matt-gardner commented:
@tingkai-zhang, @joelgrus, what's the status of this PR?

joelgrus commented:
I made a slight change to this, instead of having the default value for never split duplicate whatever the BertTokenizer uses, I made the default value None, in which case we use whatever their default value happens to be. That feels cleaner than duplicating their list.
